motionplan/tpSpaceRRT.go

```diff
@@ -362,107 +529,256 @@ func (mp *tpSpaceRRTMotionPlanner) extendMap(
+// is the distal pose away from the root of the tree, which in the case of the goal tree is in fact the 0-distance point of the traj.
+// When this becomes a single path, poses should reflect the transformation at the end of each traj. Here we go through and recompute
+// each pose in order to ensure correctness.
+// TODO: if trees are stored as segments rather than nodes, then this becomes simpler/unnecessary. Related to RSDK-4139.
+func (mp *tpSpaceRRTMotionPlanner) rectifyPath(path []node) ([]node, error) {
+	correctedPath := []node{}
+	runningPose := spatialmath.NewZeroPose()
+	for _, wp := range path {
+		wpPose, err := mp.frame.Transform(wp.Q())
+		if err != nil {
+			return nil, err
+		}
+		runningPose = spatialmath.Compose(runningPose, wpPose)
+
+		thisNode := &basicNode{
+			q:      wp.Q(),
+			cost:   wp.Cost(),
+			pose:   runningPose,
+			corner: wp.Corner(),
+		}
+		correctedPath = append(correctedPath, thisNode)
+	}
+	return correctedPath, nil
+}
+
 func (mp *tpSpaceRRTMotionPlanner) smoothPath(ctx context.Context, path []node) []node {
-	mp.logger.Info("smoothing not yet supported for tp-space")
+	toIter := int(math.Min(float64(len(path)*len(path))/2, float64(mp.planOpts.SmoothIter)))
+	currCost := sumCosts(path)
+
+	maxCost := math.Inf(-1)
+	for _, wp := range path {
+		if wp.Cost() > maxCost {
+			maxCost = wp.Cost()
+		}
+	}
+	newFrame, err := newPTGFrameFromPTGFrame(mp.frame, maxCost*mp.algOpts.smoothScaleFactor)
+	if err != nil {
+		return path
+	}
+	smoothPlannerMP, err := newTPSpaceMotionPlanner(newFrame, mp.randseed, mp.logger, mp.planOpts)
+	if err != nil {
+		return path
+	}
+	smoothPlanner := smoothPlannerMP.(*tpSpaceRRTMotionPlanner)
+	for i := 0; i < toIter; i++ {
+		select {
+		case <-ctx.Done():
+			return path
+		default:
+		}
+		// get start node of first edge. Cannot be either the last or second-to-last node.
+		// Intn will return an int in the half-open interval half-open interval [0,n)
+		firstEdge := mp.randseed.Intn(len(path) - 2)
+		secondEdge := firstEdge + 2 + mp.randseed.Intn((len(path)-2)-firstEdge)
+
+		newInputSteps, err := mp.attemptSmooth(ctx, path, firstEdge, secondEdge, smoothPlanner)
+		if err != nil || newInputSteps == nil {
+			continue
+		}
+		newCost := sumCosts(newInputSteps)
+		if newCost >= currCost {
+			continue
+		}
+		// Re-connect to the final goal
+		if newInputSteps[len(newInputSteps)-1] != path[len(path)-1] {
+			newInputSteps = append(newInputSteps, path[len(path)-1])
+		}
+
+		goalInputSteps, err := mp.attemptSmooth(ctx, newInputSteps, len(newInputSteps)-3, len(newInputSteps)-1, smoothPlanner)
+		if err != nil || goalInputSteps == nil {
+			continue
+		}
+		goalInputSteps = append(goalInputSteps, path[len(path)-1])
+		path = goalInputSteps
+		currCost = sumCosts(path)
+	}
 	return path
 }
+
+// firstEdge and secondEdge must not be adjacent.
+func (mp *tpSpaceRRTMotionPlanner) attemptSmooth(
```

Comment:
I might ask you to explain this offline, to help me understand it. A comment could also be helpful.

Reply:
Comment added